File: vision_fullcam/state/ppe_observer.py
# vision_fullcam/state/ppe_observer.py
from typing import Dict, Tuple, TYPE_CHECKING

if TYPE_CHECKING:
    from vision_fullcam.state.person_state import PersonState
    from vision_fullcam.tracking.simple_tracker import Tracked

# bbox utils (state 안에서만 쓰는 내부 로직)
def center_in(box, roi):
    x1,y1,x2,y2 = box
    cx, cy = (x1+x2)/2.0, (y1+y2)/2.0
    X1,Y1,X2,Y2 = roi
    return (X1 <= cx <= X2) and (Y1 <= cy <= Y2)

def roi_head(pb):
    x1,y1,x2,y2 = pb; h = y2-y1
    return (x1, y1, x2, int(y1+0.35*h))

def roi_torso(pb):
    x1,y1,x2,y2 = pb; h = y2-y1
    return (x1, int(y1+0.20*h), x2, int(y1+0.75*h))

def roi_feet(pb):
    x1,y1,x2,y2 = pb; h = y2-y1
    return (x1, int(y1+0.70*h), x2, y2)

def iou(a,b):
    ax1,ay1,ax2,ay2=a; bx1,by1,bx2,by2=b
    ix1,iy1=max(ax1,bx1),max(ay1,by1)
    ix2,iy2=min(ax2,bx2),min(ay2,by2)
    iw,ih=max(0,ix2-ix1),max(0,iy2-iy1)
    inter=iw*ih
    area_a=max(0,ax2-ax1)*max(0,ay2-ay1)
    area_b=max(0,bx2-bx1)*max(0,by2-by1)
    union=max(1, area_a+area_b-inter)
    return inter/union

def roi_visible(roi, frame_w, frame_h, min_area_ratio=0.3):
    x1,y1,x2,y2 = roi
    ix1,iy1=max(0,x1),max(0,y1)
    ix2,iy2=min(frame_w,x2),min(frame_h,y2)
    iw,ih=max(0,ix2-ix1),max(0,iy2-iy1)
    inter=iw*ih
    area=max(1,(x2-x1)*(y2-y1))
    return (inter/area) >= min_area_ratio


class PPEObserver:
    """
    사람별 PPE 관측 전용 클래스
    - helmet / vest / shoes
    - feet_visible (shoes gate)
    - on_ladder (사다리 탑승 여부)
    """

    def update(
        self,
        persons: Dict[int, "PersonState"],
        tracked: Dict[int, "Tracked"],
        frame_shape: Tuple[int,int],
    ):
        h, w = frame_shape

        helmet_boxes = [t.bbox for t in tracked.values() if t.label == "helmet"]
        vest_boxes   = [t.bbox for t in tracked.values() if t.label == "safety_vest"]
        shoes_boxes  = [t.bbox for t in tracked.values() if t.label == "safety_shoes"]
        ladder_boxes = [t.bbox for t in tracked.values() if t.label == "ladder"]

        for p in persons.values():
            if p.bbox is None:
                continue

            head  = roi_head(p.bbox)
            torso = roi_torso(p.bbox)
            feet  = roi_feet(p.bbox)

            has_helmet = any(center_in(b, head) for b in helmet_boxes)
            has_vest   = any(center_in(b, torso) for b in vest_boxes)

            # 발이 프레임 안에 있는지 (shoes gate)
            feet_visible = roi_visible(feet, w, h, min_area_ratio=0.35)

            if feet_visible:
                has_shoes = any(center_in(b, feet) for b in shoes_boxes)
            else:
                has_shoes = True  # 판단 유예 (오탐 방지)

            # 히스토리 업데이트
            p.helmet_hist.append(has_helmet)
            p.vest_hist.append(has_vest)
            p.feet_visible_hist.append(feet_visible)
            p.shoes_hist.append(has_shoes)

            # 사다리 탑승 여부 (feet ROI 기준)
            on_ladder = False
            for lb in ladder_boxes:
                if iou(feet, lb) > 0.10:
                    on_ladder = True
                    break
            p.on_ladder_hist.append(on_ladder)

File: frontend/mobile/src/vision/state/ppeObserver.ts
import type { BBox, Tracked } from '../types';
import type { PersonState } from './personState';

function centerIn(box: BBox, roi: BBox): boolean {
  const cx = (box[0] + box[2]) / 2;
  const cy = (box[1] + box[3]) / 2;
  return roi[0] <= cx && cx <= roi[2] && roi[1] <= cy && cy <= roi[3];
}

function roiHead(pb: BBox): BBox {
  const h = pb[3] - pb[1];
  return [pb[0], pb[1], pb[2], Math.round(pb[1] + 0.35 * h)];
}

function roiTorso(pb: BBox): BBox {
  const h = pb[3] - pb[1];
  return [pb[0], Math.round(pb[1] + 0.20 * h), pb[2], Math.round(pb[1] + 0.75 * h)];
}

function roiFeet(pb: BBox): BBox {
  const h = pb[3] - pb[1];
  return [pb[0], Math.round(pb[1] + 0.70 * h), pb[2], pb[3]];
}

function iou(a: BBox, b: BBox): number {
  const ix1 = Math.max(a[0], b[0]);
  const iy1 = Math.max(a[1], b[1]);
  const ix2 = Math.min(a[2], b[2]);
  const iy2 = Math.min(a[3], b[3]);
  const iw = Math.max(0, ix2 - ix1);
  const ih = Math.max(0, iy2 - iy1);
  const inter = iw * ih;
  const areaA = Math.max(0, a[2] - a[0]) * Math.max(0, a[3] - a[1]);
  const areaB = Math.max(0, b[2] - b[0]) * Math.max(0, b[3] - b[1]);
  const union = Math.max(1, areaA + areaB - inter);
  return inter / union;
}

function roiVisible(
  roi: BBox,
  frameW: number,
  frameH: number,
  minAreaRatio: number = 0.35,
): boolean {
  const ix1 = Math.max(0, roi[0]);
  const iy1 = Math.max(0, roi[1]);
  const ix2 = Math.min(frameW, roi[2]);
  const iy2 = Math.min(frameH, roi[3]);
  const iw = Math.max(0, ix2 - ix1);
  const ih = Math.max(0, iy2 - iy1);
  const inter = iw * ih;
  const area = Math.max(1, (roi[2] - roi[0]) * (roi[3] - roi[1]));
  return inter / area >= minAreaRatio;
}

export class PPEObserver {
  update(
    persons: Map<number, PersonState>,
    tracked: Map<number, Tracked>,
    frameShape: [number, number], // [height, width]
  ): void {
    const [h, w] = frameShape;

    const helmetBoxes: BBox[] = [];
    const vestBoxes: BBox[] = [];
    const shoesBoxes: BBox[] = [];
    const ladderBoxes: BBox[] = [];

    for (const t of tracked.values()) {
      if (t.label === 'helmet') helmetBoxes.push(t.bbox);
      else if (t.label === 'safety_vest') vestBoxes.push(t.bbox);
      else if (t.label === 'safety_shoes') shoesBoxes.push(t.bbox);
      else if (t.label === 'ladder') ladderBoxes.push(t.bbox);
    }

    for (const p of persons.values()) {
      if (p.bbox === null) continue;

      const head = roiHead(p.bbox);
      const torso = roiTorso(p.bbox);
      const feet = roiFeet(p.bbox);

      const hasHelmet = helmetBoxes.some(b => centerIn(b, head));
      const hasVest = vestBoxes.some(b => centerIn(b, torso));

      const feetVisible = roiVisible(feet, w, h, 0.35);

      let hasShoes: boolean;
      if (feetVisible) {
        hasShoes = shoesBoxes.some(b => centerIn(b, feet));
      } else {
        hasShoes = true; // 판단 유예 (오탐 방지)
      }

      p.helmetHist.push(hasHelmet);
      p.vestHist.push(hasVest);
      p.feetVisibleHist.push(feetVisible);
      p.shoesHist.push(hasShoes);

      console.log(
        `[Vision:PPE] person#${p.id} helmet=${hasHelmet} vest=${hasVest} shoes=${hasShoes} feetVisible=${feetVisible}`,
      );

      // 사다리 탑승 여부
      let onLadder = false;
      for (const lb of ladderBoxes) {
        if (iou(feet, lb) > 0.10) {
          onLadder = true;
          break;
        }
      }
      p.onLadderHist.push(onLadder);
    }
  }
}
